plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -524,24 +571,32 @@ applyEvaluate _ val _ = throwingDischarged _MachineError NonFunctionalApplicatio
 
 -- | Apply a builtin to a list of CekValue arguments
 applyBuiltin
-    :: Ix fun
+    :: forall cost uni fun s
+    . (Ix fun, PrettyUni uni fun)
     => Context uni fun
     -> fun
     -> [CekValue uni fun]
     -> CekM cost uni fun s (Term Name uni fun ())
 applyBuiltin ctx bn args = do
-  -- Turn the cause of a possible failure, being a 'CekValue', into a 'Term'.
+
+  rt <- asks cekEnvRuntime
+  BuiltinRuntime sch _ f exF <- lookupBuiltinEx (Proxy @(CekEvaluationException uni fun)) bn rt
+
+  -- ''applyTypeSchemed' doesn't throw exceptions so that we can easily catch them here and
+  -- post-process them.
   -- See Note [Being generic over @term@ in 'CekM'].
-  let dischargeError = hoist $ withExceptT $ mapCauseInMachineException $ void . dischargeCekValue
-  BuiltinRuntime sch _ f exF <- asksM $ lookupBuiltin bn . cekEnvRuntime
-  result <- dischargeError $ applyTypeSchemed bn sch f exF args
-  returnCek ctx result
+  resultOrErr <- runExceptT $ applyTypeSchemed bn sch f exF args
+  case resultOrErr of
+      -- Turn the cause of a possible failure, being a 'CekValue', into a 'Term'.
+      -- See Note [Being generic over @term@ in 'CekM'].
+      Left e       -> throwCek $ mapCauseInMachineException (void . dischargeCekValue) e
```

Comment:
Well, there's that issue that we shouldn't discharge values and that probably applies to exceptions as well, so it seems we should do something different here?

Reply:
Yeah, but I'm not worrying about that right now.